File: DSA-with-Python/SinglyLinkedList/sllist.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None
        
        
class List:
    def __init__(self):
        self.size = 0
        self.head = None

    def push(self, data):
        u = Node(data)
        u.next = self.head
        self.head = u
        
        self.size+=1

    def pop(self):
        if self.size==0:
            return 0
        u = self.head
        self.head = self.head.next
        
        self.size-=1
        
        return u.data
    
    def printData(self):
        u = self.head
        while u is not None:
            print(u.data)
            u = u.next    
    

def main():
    ob = List()
    
    ob.push(5)
    ob.push(6)
    ob.push(7)
    
    
    print(ob.pop())
    print(ob.pop())
    print('==')
    ob.printData()
    
    
if __name__=="__main__":
    main() 

File: SinglyLinkedList/sllist.cc
/**
 * ** Prerequisite to understand this code
 *  -> Object Oriented Programming
*/
#include <iostream>
using namespace std;

struct Node{
  int dta;      //stores data
  Node* nxt;    //stores pointers of next node
};

class List{
private:
  Node* head;    //stores pointers of first node
  Node* tail;    //stores pointers of last node
  int n;

  //At first we need to set a program that can create nodes
  Node* newNode(int x){
    Node* node = new Node;
    node->dta = x;
    node->nxt = nullptr;

    return node;

    //if you want to write 'newNode' in one line then
    //return new Node{x, nullptr};
  }

public:
  List();    //Constructor;

  //Stack interface
  void Push(int x);
  int Pop();

  //Dequeqe interface
  void Add(int x);
  int Remove();

  void PrintList();    //For printing the List 
};

List::List(){
  //Initializing variables;
  n = 0;
  head = nullptr;
  tail = nullptr;
}


//*************STACK INTERFACE****************
void List::Push(int x){
  Node* u = newNode(x);
  u->nxt = head;
  head = u;
  if(n==0) tail = u;
  ++n;
  return;    //As it is a void function;
}

int List::Pop(){
  if(n==0) return 0;   //No values to remove;
  int temp = head->dta;
  head = head->nxt;
  --n;
  return temp;
}


//************QUEUE INTERFACE***************
void List::Add(int x){
  Node* u = newNode(x);
  if(n==0) head = u;
  else tail->nxt = u;
  tail = u;
  ++n;
  return;
}

int List::Remove(){
  return Pop();    //Same as pop()
}


//*************PRINTING****************
void List::PrintList(){
  Node* u = head;
  cout<<"[ ";
  while(u!=nullptr){
    cout<<u->dta<<' ';
    u = u->nxt;
  }
  cout<<"]\n";
  return;
}
int main() {
  List ob;
  ob.Push(5);
  ob.Push(4);
  ob.Push(3);
  ob.Push(2);
  ob.Push(1);

  ob.Add(5);
  ob.Add(4);
  ob.Add(3);
  ob.Add(2);
  ob.Add(1);

  // ob.Pop();
  // ob.Pop();
  // ob.Pop();
  // ob.Pop();
  // ob.Pop();
  // ob.Pop();

  // ob.Remove();
  // ob.Remove();
  // ob.Remove();
  // ob.Remove();
  // ob.Remove();
  // ob.Remove();
  
  ob.PrintList();
  return 0;
}

//Comment If you find any mistake;